File: cpp/tests/src/tests/Stop.cpp
/*
 * Copyright 2015 Institut Laue-Langevin
 *
 * Licensed under the EUPL, Version 1.1 only (the "License");
 * You may not use this work except in compliance with the Licence.
 * You may obtain a copy of the Licence at:
 *
 * http://joinup.ec.europa.eu/software/page/eupl
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the Licence is distributed on an "AS IS" basis,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the Licence for the specific language governing permissions and
 * limitations under the Licence.
 */

#include <cameo/api/cameo.h>
#include <iostream>

using namespace std;
using namespace cameo;

struct StopState {
	bool stopping;

	StopState() : stopping(false) {}
};

struct Stop {

	shared_ptr<StopState> data;

	Stop(shared_ptr<StopState> sharedData) : data(sharedData) {}

	void operator()() {
		cout << "Stop handler executed" << endl;
		data->stopping = true;
	}
};

int main(int argc, char *argv[]) {

	application::This::init(argc, argv);

	// New block to ensure cameo objects are terminated before the application.
	{
		application::This::setRunning();

		// Define an object StopState that is shared with the handler.
		shared_ptr<StopState> state(new StopState());
		application::This::handleStop(Stop(state));

		int i = 0;
		while (!state->stopping) {
			cout << "Waiting " << i << "..." << endl;
			this_thread::sleep_for(chrono::milliseconds(100));
			i++;
		}

		application::This::setResult("2189");

		cout << "Finished the application" << endl;
	}

	return 0;
}

File: java/tests/src/fr/ill/ics/cameo/test/Stop.java
/*
 * Copyright 2015 Institut Laue-Langevin
 *
 * Licensed under the EUPL, Version 1.1 only (the "License");
 * You may not use this work except in compliance with the Licence.
 * You may obtain a copy of the Licence at:
 *
 * http://joinup.ec.europa.eu/software/page/eupl
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the Licence is distributed on an "AS IS" basis,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the Licence for the specific language governing permissions and
 * limitations under the Licence.
 */

package fr.ill.ics.cameo.test;

import java.util.Date;

import fr.ill.ics.cameo.base.Application;
import fr.ill.ics.cameo.base.This;


public class Stop {

	public static boolean stopping = false;
	
	public static void main(String[] args) {

		This.init(args);
		
		try {
			Date begin = new Date();
			
			This.handleStop(new Application.Handler() {
				
				@Override
				public void handle() {
					stopping = true;
					System.out.println("Stop handler executed");
				}
			});
			
			int i = 0;
			while (!stopping) {
				System.out.println("Waiting " + i + "...");
				Thread.sleep(100);
				i++;
			}
			
			Date end = new Date();
			
			String result = "";
			result += end.getTime() - begin.getTime();
			
			This.setResult(result);
			
			System.out.println("Finished the application");
		}
		catch (Throwable e) {
			e.printStackTrace();
		}
		finally {
			This.terminate();			
		}
	}

}